--- a/Makeconfig
+++ b/Makeconfig
@@ -153,6 +153,11 @@
 libdir = $(exec_prefix)/lib
 endif
 inst_libdir = $(install_root)$(libdir)
+
+# Compat places to look for libraries
+ifndef extra_libdir
+extra_libdir = /lib:$(exec_prefix)/lib
+endif
 
 # Where to install the shared library.
 ifndef slibdir
@@ -661,6 +666,10 @@
 default-rpath = $(libdir)
 endif
 
+ifdef extra_libdir
+default-rpath += :$(extra_libdir)
+endif
+
 ifndef link-extra-libs
 link-extra-libs = $(LDLIBS-$(@F))
 link-extra-libs-static = $(link-extra-libs)
--- a/elf/Makefile
+++ b/elf/Makefile
@@ -1690,7 +1690,7 @@
 	$(make-target-directory)
 	echo "$(subst :, ,$(default-rpath) $(user-defined-trusted-dirs))"    \
 	| $(AWK) -f gen-trusted-dirs.awk > ${@:st=T};
-	echo '#define DL_DST_LIB "$(notdir $(slibdir))"' >> ${@:st=T}
+	echo '#define DL_DST_LIB "$(shell echo $(slibdir) | sed 's,/,,')"' >> ${@:st=T}
 	$(move-if-change) ${@:st=T} ${@:st=h}
 	touch $@
 CPPFLAGS-dl-load.c += -I$(objpfx). -I$(csu-objpfx).
--- a/elf/ldconfig.c
+++ b/elf/ldconfig.c
@@ -59,6 +59,17 @@
 #include <version.h>
 
 #define PACKAGE _libc_intl_domainname
+
+/* Get the generated information about the trusted/standard directories.  */
+#include "trusted-dirs.h"
+
+static const char system_dirs[] = SYSTEM_DIRS;
+static const size_t system_dirs_len[] =
+{
+  SYSTEM_DIRS_LEN
+};
+#define nsystem_dirs_len \
+  (sizeof (system_dirs_len) / sizeof (system_dirs_len[0]))
 
 /* List of directories to handle.  */
 struct dir_entry
@@ -480,6 +491,25 @@
   return ret;
 }
 
+static const char * const ld_sonames[] =
+{
+  "ld-kfreebsd-x86-64.so.1",
+  "ld-linux-aarch64.so.1",
+  "ld-linux-aarch64_be.so.1",
+  "ld-linux-armhf.so.3",
+  "ld-linux-ia64.so.2",
+  "ld-linux-mipsn8.so.1",
+  "ld-linux-riscv64-lp64.so.1"
+  "ld-linux-riscv64-lp64d.so.1"
+  "ld-linux-x32.so.2",
+  "ld-linux-x86-64.so.2",
+  "ld-linux.so.2",
+  "ld-linux.so.3",
+  "ld.so.1",
+  "ld64.so.1",
+  "ld64.so.2",
+};
+
 /* Create a symbolic link from soname to libname in directory path.  */
 static void
 create_links (const char *real_path, const char *path, const char *libname,
@@ -490,6 +520,7 @@
   struct stat stat_lib, stat_so, lstat_so;
   int do_link = 1;
   int do_remove = 1;
+  int i;
   /* XXX: The logics in this function should be simplified.  */
 
   /* Get complete path.  */
@@ -518,6 +549,18 @@
 	  error (0, 0, _("Can't stat %s\n"), full_libname);
 	  return;
 	}
+
+      /* Do not change the symlink pointer to the dynamic linker except for
+	 non-existing symlinks, as it might break multiarch systems.  */
+      for (i = 0; i < sizeof (ld_sonames) / sizeof (ld_sonames[0]); i++)
+	if (__glibc_unlikely(!strcmp(soname, ld_sonames[i])))
+	  {
+	    if (opt_verbose)
+	      error (0, 0, _("%s is the dynamic linker, ignoring\n"),
+			     full_libname);
+	    do_link = 0;
+	  }
+
       if (stat_lib.st_dev == stat_so.st_dev
 	  && stat_lib.st_ino == stat_so.st_ino)
 	/* Link is already correct.  */
@@ -1324,9 +1367,15 @@
       ldconfig_parse_config (config_file, opt_chroot, add_dir_callback);
 
       /* Always add the standard search paths.  */
-      add_system_dir (SLIBDIR);
-      if (strcmp (SLIBDIR, LIBDIR))
-	add_system_dir (LIBDIR);
+      const char *strp = system_dirs;
+      size_t idx = 0;
+      do
+        {
+          add_system_dir (strp);
+          strp += system_dirs_len[idx] + 1;
+          idx++;
+        }
+      while (idx < nsystem_dirs_len);
     }
 
   const char *aux_cache_file = _PATH_LDCONFIG_AUX_CACHE;
